parser/printer.go

```diff
@@ -330,36 +330,60 @@ func (p *printer) flush() {
 		p.curComment++
 	}
 	p.output = append(p.output, '\n')
+// a VerbosePrinter prints a syntax tree and is intended for debugging purposes
+type VerbosePrinter struct {
+	syntaxTree *SyntaxTree
+	output     bytes.Buffer
+	indent     int
 }
 
-// Print whitespace to pad from column l to column max
-func (p *printer) pad(l int) {
-	if l > len(p.wsBuf) {
-		p.wsBuf = make([]byte, l)
-		for i := range p.wsBuf {
-			p.wsBuf[i] = ' '
-		}
-	}
-	p.output = append(p.output, p.wsBuf[0:l]...)
+func VerbosePrint(tree *SyntaxTree) string {
+	return NewVerbosePrinter(tree).Print()
 }
 
-func (p *printer) indent(i int) {
-	p.indentList = append(p.indentList, i)
+func NewVerbosePrinter(tree *SyntaxTree) (p *VerbosePrinter) {
+	return &VerbosePrinter{tree, bytes.Buffer{}, 0}
 }
 
-func (p *printer) unindent(pos scanner.Position) {
-	p.printEndOfLineCommentsBefore(pos)
-	p.indentList = p.indentList[0 : len(p.indentList)-1]
+func (p *VerbosePrinter) Print() string {
```

Comment:
This is the only public method, and always used as NewVerbosePrinter(...).Print(), might as well make a VerbosePrint(...) method that delegates to an unexported verbosePrinter type.